core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dc.append(l_c)
+            for w, l in zip(doc, l_c):
+                self._update_counting_matrices(d, w, l, 1)
+
+    def _infer_docs(self, doc_ids):
+        """Runs iterative inference on given docs (default all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for c in xrange(len(doc)):
+                w = doc[c]
+                l = self._l_dc[d][c]
+
+                self._update_counting_matrices(d, w, l, -1)
+
+                # Gibbs update of labels
+                coeff_a = 1 / (
+                    self._c_dl[d].sum() + self._label_count * self._alpha)
+                coeff_b = 1 / (
+                    self._c_lw.sum(axis=1) + self._word_count * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
+                    1, prob_l / prob_l.sum()
+                    ).argmax()
+
+                statez['computes'] += 1
+                if l != new_l:
+                    statez['updates'] += 1
+
+                self._l_dc[d][c] = new_l
+                self._update_counting_matrices(d, w, new_l, 1)
+
+        return statez
+
+    def _get_probabilities(self, d):
+        """Returns the probability of a document having a label."""
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probabilities_with_label(self, d):
+        """Includes human readable labels with document probabilities."""
+        probs_i = self._get_probabilities(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc"""
+        probs_l = self._get_probabilities_with_label(d)
+        default_prob = probs_l['_default']
+        for l in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[l]
+            if l != '_default' and prob_l / (1 - default_prob) > threshold:
+                return l, prob_l / (1 - default_prob), probs_l
+        return '_default', default_prob / (1 - default_prob), probs_l
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
```

Comment:
well, I wanted to place as little restrictions on a label as possible, since any string can be a label. This gives more independence to the client. I only removed using underscores initially in case we want internal labels in the future beyond _default.

Do you think it would be beneficial to have more restrictions?